internal/peer/common/common.go

```diff
@@ -305,6 +305,13 @@ func InitCmd(cmd *cobra.Command, args []string) {
 		LogSpec: loggingSpec,
 	})
 
+	// 'peer lifecycle chaincode package' mandates does not need to init crypto
+	// For details, please refer to FAB-17453
+	if cmd != nil && strings.Contains(cmd.CommandPath(), "lifecycle") && cmd.Name() == "package" {
```

Comment:
Checking to see if the command path contains the word "lifecycle" is too fragile, IMO. If you can write a debug message the explicitly states the path, we can certainly use that as the check.

Additionally, why are we checking for a nil cmd? If we get a nil command, something is semantically wrong. How are we to initialize a command when there is no command? Why are we initializing a command if there is no command?

It may be necessary given the existing (poor) structure of the code but if it is, it would be nice to fix the source of that issue as well.

Reply:
Cmd cannot be empty in real environment, and the nil cmd only appear in test functions to simplify parameters, as TestInitCmd function in common_test.go. So I check if it is nil.

Reply:
The root cause of this problem is that chaincodeCmd(in file internal/peer/lifecycle/chaincode/chaincode.go) registers the common.InitCmd function as a hook function, which has package, install, commit and other subcommads, but the cobra framework does not provide parameters to suppress this hook function for a subcommand. So we cannot treat these subcommands differently at this stage.
Define a new cobra.Command variable to register "package" command, which will not register function common.InitCmd as a hook function, instead of  chaincodeCmd. This approach may be better, but it will have some impact on the existing command line format,  which may require more community proposals. So I chose this tactical fix.

Reply:
> The root cause of this problem is that chaincodeCmd(in file internal/peer/lifecycle/chaincode/chaincode.go) registers the common.InitCmd function as a hook function

I understand the cause and I am familiar with this code. A larger non-tactical fix is certainly possible given the InitCmd calls are coming from the pre-execution hooks in the commands we register. Regardless, that wasn’t one of the things holding up this merge.

> Cmd cannot be empty in real environment, and the nil cmd only appear in test functions to simplify parameters, as TestInitCmd function in common_test.go. So I check if it is nil.

It’s in exactly _one_ test that we pass a nil. The test needs to be changed to pass an empty command instead of nil and the production code updated such that it does not check for a nil command.

> Checking to see if the command path contains the word "lifecycle" is too fragile

The point I was trying to make with that comment was that the production code change for a tactical fix like this should be as simple as:
```go
    if cmd.CommandPath() == "peer lifecycle chaincode package" {
        mainLogger.Debug("peer lifecycle chaincode package does not need to init crypto")
        return
    }
```

With that one block in the production code, your test would need to change to provide a “peer” command in your setup and the TestInitCmd test would have to pass `&cobra.Command{}` as the first argument to `InitCmd`. With that, the fix and tests would be complete.

Reply:
My understanding was a bit off before, I'm very sorry. It has been modified again according to your suggestion, please review.